fuel/datasets/mnist.py

```diff
@@ -45,50 +39,47 @@ class MNIST(InMemoryDataset):
     binary : bool, optional
         If ``True``, returns binary (black/white) images instead of
         grayscale. ``False`` by default.
+    flatten : bool, optional
+        If ``True``, returns flattened MNIST examples i.e. vectors of
+        length 784 instead of 28 x 28 images. Defaults to ``True``.
 
     """
     provides_sources = ('features', 'targets')
-
-    def __init__(self, which_set, start=None, stop=None, binary=False,
-                 **kwargs):
+    folder = 'mnist'
+    files = {
+        'train': {'images': 'train-images-idx3-ubyte',
+                  'labels': 'train-labels-idx1-ubyte'},
+        'test': {'images': 't10k-images-idx3-ubyte',
+                 'labels': 't10k-labels-idx1-ubyte'}
+    }
+
+    def __init__(self, which_set, binary=False, flatten=True, **kwargs):
         if which_set not in ('train', 'test'):
             raise ValueError("MNIST only has a train and test set")
-        if stop is None:
-            stop = 60000 if which_set == "train" else 10000
-        if start is None:
-            start = 0
-        self.num_examples = stop - start
-        self.default_scheme = SequentialScheme(self.num_examples, 1)
-        super(MNIST, self).__init__(**kwargs)
-
         self.which_set = which_set
-        self.start = start
-        self.stop = stop
         self.binary = binary
+        self.flatten = flatten
+
+        super(MNIST, self).__init__(OrderedDict(zip(self.provides_sources,
+                                                    self._load_mnist())),
+                                    **kwargs)
 
     def load(self):
-        if self.which_set == 'train':
-            data = 'train-images-idx3-ubyte'
-            labels = 'train-labels-idx1-ubyte'
-        elif self.which_set == 'test':
-            data = 't10k-images-idx3-ubyte'
-            labels = 't10k-labels-idx1-ubyte'
-        data_path = os.path.join(config.data_path, 'mnist')
-        x = read_mnist_images(
-            os.path.join(data_path, data),
-            'bool' if self.binary else config.floatX)[self.start:self.stop]
-        x = x.reshape((x.shape[0], numpy.prod(x.shape[1:])))
-        y = read_mnist_labels(
-            os.path.join(data_path, labels))[self.start:self.stop,
-                                             numpy.newaxis]
-        self.features = x
-        self.targets = y
-
-    def get_data(self, state=None, request=None):
-        if state is not None:
-            raise ValueError("MNIST does not have a state")
-        return self.filter_sources((self.features[request],
-                                    self.targets[request]))
+        self.indexables = [data[self.start:self.stop] for source, data
```

Comment:
It shouldn't go in `IndexableDataset`; that class is too general IMHO. `IndexableDataset` can also just be used with things like `IndexableDataset(range(100))`, in which case a default implementation of load doesn't make sense. I'm also afraid that a user might subclass it, just add a lazy property (which doesn't necessarily have to be `indexables`) and suddenly gets errors because it's trying to call a method that doesn't exist.

I do agree that this could probably be filtered out, but there'd need to be another base class, `IndexableFromDisk` or something like that maybe?